solr/core/src/java/org/apache/solr/handler/IndexFetcher.java

```diff
@@ -1988,16 +1989,38 @@ private FastInputStream getStream() throws IOException {
         req.setBasePath(leaderBaseUrl);
         if (useExternalCompression) req.addHeader("Accept-Encoding", "gzip");
         response = solrClient.request(req, leaderCoreName);
+        final var responseStatus = (Integer) response.get("responseStatus");
         is = (InputStream) response.get("stream");
+
+        if (responseStatus != 200) {
+          final var errorMsg =
+              String.format(
+                  Locale.ROOT,
+                  "Unexpected status code [%d] when downloading file [%s].",
+                  responseStatus,
+                  fileName);
+          closeStreamAndThrowIOE(is, errorMsg, Optional.empty());
+        }
+
         if (useInternalCompression) {
           is = new InflaterInputStream(is);
         }
         return new FastInputStream(is);
       } catch (Exception e) {
-        // close stream on error
-        IOUtils.closeQuietly(is);
-        throw new IOException("Could not download file '" + fileName + "'", e);
+        closeStreamAndThrowIOE(is, "Could not download file '" + fileName + "'", Optional.of(e));
+      }
+
+      // Unreachable b/c closeStreamAndThrowIOE will always throw
+      return null;
+    }
+
+    private void closeStreamAndThrowIOE(
+        InputStream is, String exceptionMessage, Optional<Exception> e) throws IOException {
```

Comment:
A matter of taste but I would prefer we/programmers use Optionals conservatively -- when it's clearly a help.  They are verbose; should have been built into the language like Kotlin.  In particular, I suggest we not use Optional as a parameter; only return them if we use them.

See https://www.baeldung.com/java-optional #15 (and I've seen this advice elsewhere too)

Reply:
We really need more coding style guides and ideally lint checking tools for things like Optional.   Or in 12 months, we'll have this same conversation, but with someone new ;-(.